src/prefect/deployments.py

```diff
@@ -172,3 +240,195 @@ def cast_storage_to_block_type(cls, value):
+                    if deployment.storage_document_id:
+                        self.storage = Block._from_block_document(
+                            await client.read_block_document(
+                                deployment.storage_document_id
+                            )
+                        )
+            except ObjectNotFound:
+                return False
+        return True
+
+    @sync_compatible
+    async def update(self, ignore_none: bool = False, **kwargs):
+        """
+        Performs an in-place update with the provided settings.
+        """
+        unknown_keys = set(kwargs.keys()) - set(self.dict().keys())
+        if unknown_keys:
+            raise ValueError(
+                f"Received unexpected attributes: {', '.join(unknown_keys)}"
+            )
+        for key, value in kwargs.items():
+            if ignore_none and value is None:
+                continue
+            setattr(self, key, value)
+
+    @sync_compatible
+    async def upload_to_storage(self, storage_block: Block = None) -> Optional[int]:
+        """
+        Uploads the workflow this deployment represents using a provided storage block;
+        if no block is provided, defaults to configuring self for local storage.
+        """
+        deployment_path = None
+        file_count = None
+        if storage_block:
+            template = await Block.load(storage_block)
+            self.storage = template.copy(
+                exclude={"_block_document_id", "_block_document_name", "_is_anonymous"}
+            )
+
+            # upload current directory to storage location
+            file_count = await self.storage.put_directory(ignore_file=".prefectignore")
+        elif not self.storage:
+            # default storage, no need to move anything around
+            self.storage = None
+            deployment_path = str(Path(".").absolute())
+        else:
+            file_count = await self.storage.put_directory(ignore_file=".prefectignore")
+
+        # persists storage now in case it contains secret values
+        if self.storage and not self.storage._block_document_id:
+            await self.storage._save(is_anonymous=True)
+
+        self.path = deployment_path
+        return file_count
+
+    @sync_compatible
+    async def to_yaml(self, output: str):
+        """
+        Compiles the current deployment into a YAML file at the location specified by `output`.
+        """
+        with open(output, "w") as f:
+            f.write(self.header)
+            yaml.dump(self._editable_fields_dict(), f, sort_keys=False)
+            do_not_edit_msg = " DO NOT EDIT BELOW THIS LINE "
+            msg_length = len(do_not_edit_msg)
+            f.write(
+                f"###{' ' * msg_length}###\n###{do_not_edit_msg}###\n###{' ' * msg_length}###\n"
+            )
+            yaml.dump(self._immutable_fields_dict(), f, sort_keys=False)
+
+    @sync_compatible
+    async def apply(self) -> UUID:
```

Comment:
general developer experience question - how careful should we be when using the `sync_compatible` decorator for common user interfaces?

IIRC the snag is anything decorated with `sync_compatible` will need to be awaited ONLY IF it is called in an async context

This might cause some weirdness for users. I think this might have already caused a bug during internal development as well?

Reply:
Oh boy I have a lot of thoughts on this -- the way I thought about it in this case was that this class and its methods will generally be called within a standalone python script, and most likely in a `if __name__ == "__main__":` block.  Starting up your own event loop is a whole thing in this situation, so most likely users will want to call it synchronously.  

I also had a bug where I called a sync compatible method within an async function and the issue was subtle - I wonder if we could find a way to raise if these aren't awaited or something